gpbft/emulator/emulator.go

```diff
@@ -0,0 +1,93 @@
+package emulator
+
+import (
+	"fmt"
+
+	"github.com/filecoin-project/go-f3/gpbft"
+)
+
+// Emulator emulates the core GPBFT protocol for one honest participant, and
+// allows frame-by-frame control over progress of a GPBFT instance.
+//
+// Progress may be made by calling Emulator.ReceiveMessage,
+// Emulator.ReceiveBroadcast or Emulator.ReceiveAlarm.
+type Emulator struct {
+	subject *gpbft.Participant
+	host    *host
+}
+
+// New instantiates a new Emulator with the given GPBFT options. See
+// Emulator.Start.
+func New(o ...gpbft.Option) (*Emulator, error) {
+	h := newHost()
+	participant, err := gpbft.NewParticipant(h, o...)
+	if err != nil {
+		return nil, err
+	}
+	return &Emulator{
+		subject: participant,
+		host:    h,
+	}, nil
+}
+
+// Start starts emulation for the given instance and signals the start of
+// instance to the emulated honest gpbft.Participant.
+//
+// See NewInstance.
+func (e *Emulator) Start(instance *Instance) error {
+	if err := e.host.setInstance(instance); err != nil {
+		return err
+	}
+	return e.subject.StartInstance(instance.id)
+}
+
+// ReceiveAlarm triggers any pending alarms. Returns true if the alarm is
+// triggered along with any errors that may occur as a result of triggering the
+// alarm. Otherwise returns false with no error.
+func (e *Emulator) ReceiveAlarm() (bool, error) {
+	if e.host.maybeTriggerAlarm() {
+		return true, e.subject.ReceiveAlarm()
+	}
+	return false, nil
+}
+
+// ReceiveMessage delivers a message to the emulated participant from the given
+// sender. The sender must not be the ID of the honest participant as
+// impersonation of messages to the participant is not supported.
```

Comment:
> Why this restriction?

My thought was to avoid emulating unrealistic scenarios, where a tester (or attacker if you like) can sign messages on behalf of the honest participant. We might want to test message replay, but I think that's different from impersonation.

> drop this, probably by removing `host.id` entirely.

Although it is true that a participant does not know its own ID, but the concept of ID is (for better or worse) present across the core implementation, used as a way to index received messages, etc. 

It is also the way the Power Table is indexed. From the emulation's perspective we want to be able to send a message on behalf of a node with specific storage power. Hence the presence of ID.